File: win-prog/ProgWinVB/Beziers and Other Splines/CanonicalSpline/CanonicalSpline.vb
'------------------------------------------------
' CanonicalSpline.vb (c) 2002 by Charles Petzold
'------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class CanonicalSpline
    Inherits Form

    Protected apt(3) As Point
    Protected fTension As Single = 0.5

    Shared Sub Main()
        Application.Run(New CanonicalSpline())
    End Sub

    Sub New()
        Text = "Canonical Spline"
        BackColor = SystemColors.Window
        ForeColor = SystemColors.WindowText
        ResizeRedraw = True

        Dim scrbar As New VScrollBar()
        scrbar.Parent = Me
        scrbar.Dock = DockStyle.Right
        scrbar.Minimum = -100
        scrbar.Maximum = 109
        scrbar.SmallChange = 1
        scrbar.LargeChange = 10
        scrbar.Value = CInt(10 * fTension)
        AddHandler scrbar.ValueChanged, AddressOf ScrollOnValueChanged

        OnResize(EventArgs.Empty)
    End Sub

    Private Sub ScrollOnValueChanged(ByVal obj As Object, _
                                     ByVal ea As EventArgs)
        Dim scrbar As ScrollBar = DirectCast(obj, ScrollBar)
        fTension = scrbar.Value / 10.0F
        Invalidate(False)
    End Sub

    Protected Overrides Sub OnResize(ByVal ea As EventArgs)
        MyBase.OnResize(ea)

        Dim cx As Integer = ClientSize.Width
        Dim cy As Integer = ClientSize.Height

        apt(0) = New Point(cx \ 4, cy \ 2)
        apt(1) = New Point(cx \ 2, cy \ 4)
        apt(2) = New Point(cx \ 2, 3 * cy \ 4)
        apt(3) = New Point(3 * cx \ 4, cy \ 2)
    End Sub

    Protected Overrides Sub OnMouseDown(ByVal mea As MouseEventArgs)
        Dim pt As Point

        If mea.Button = MouseButtons.Left Then
            If ModifierKeys = Keys.Shift Then
                pt = apt(0)
            ElseIf ModifierKeys = Keys.None Then
                pt = apt(1)
            Else
                Return
            End If
        ElseIf mea.Button = MouseButtons.Right Then
            If ModifierKeys = Keys.None Then
                pt = apt(2)
            ElseIf ModifierKeys = Keys.Shift Then
                pt = apt(3)
            Else
                Return
            End If
        Else
            Return
        End If

        Cursor.Position = PointToScreen(pt)
    End Sub

    Protected Overrides Sub OnMouseMove(ByVal mea As MouseEventArgs)
        Dim pt As New Point(mea.X, mea.Y)

        If mea.Button = MouseButtons.Left Then
            If ModifierKeys = Keys.Shift Then
                apt(0) = pt
            ElseIf ModifierKeys = Keys.None Then
                apt(1) = pt
            Else
                Return
            End If
        ElseIf mea.Button = MouseButtons.Right Then
            If ModifierKeys = Keys.None Then
                apt(2) = pt
            ElseIf ModifierKeys = Keys.Shift Then
                apt(3) = pt
            Else
                Return
            End If
        Else
            Return
        End If

        Invalidate()
    End Sub

    Protected Overrides Sub OnPaint(ByVal pea As PaintEventArgs)
        Dim grfx As Graphics = pea.Graphics
        Dim br As New SolidBrush(ForeColor)
        Dim i As Integer

        grfx.DrawCurve(New Pen(ForeColor), apt, fTension)
        grfx.DrawString("Tension = " & fTension.ToString(), Font, br, 0, 0)

        For i = 0 To 3
            grfx.FillEllipse(br, apt(i).X - 3, apt(i).Y - 3, 7, 7)
        Next i
    End Sub
End Class

File: win-prog/ProgWinCS/Beziers and Other Splines/CanonicalSpline/CanonicalSpline.cs
//----------------------------------------------
// CanonicalSpline.cs © 2001 by Charles Petzold
//----------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class CanonicalSpline: Form
{
     protected Point[] apt      = new Point[4];
     protected float   fTension = 0.5f;

     public static void Main()
     {
          Application.Run(new CanonicalSpline());
     }
     public CanonicalSpline()
     {
          Text = "Canonical Spline";
          BackColor = SystemColors.Window;
          ForeColor = SystemColors.WindowText;
          ResizeRedraw = true;

          ScrollBar scroll = new VScrollBar();
          scroll.Parent = this;
          scroll.Dock = DockStyle.Right;
          scroll.Minimum = -100;
          scroll.Maximum =  109;
          scroll.SmallChange =  1;
          scroll.LargeChange = 10;
          scroll.Value = (int) (10 * fTension);
          scroll.ValueChanged += new EventHandler(ScrollOnValueChanged);

		  OnResize(EventArgs.Empty);
     }
     void ScrollOnValueChanged(object obj, EventArgs ea)
     {
          ScrollBar scroll = (ScrollBar) obj;

          fTension = scroll.Value / 10f;

          Invalidate(false);
     }
     protected override void OnResize(EventArgs ea)
     {
          base.OnResize(ea);

          int cx = ClientSize.Width;
          int cy = ClientSize.Height;

          apt[0] = new Point(    cx / 4,     cy / 2);
          apt[1] = new Point(    cx / 2,     cy / 4);
          apt[2] = new Point(    cx / 2, 3 * cy / 4);
          apt[3] = new Point(3 * cx / 4,     cy / 2);
     }
     protected override void OnMouseDown(MouseEventArgs mea)
     {
          Point pt;

          if (mea.Button == MouseButtons.Left)
          {
               if (ModifierKeys == Keys.Shift)
                    pt = apt[0];
               else if (ModifierKeys == Keys.None)
                    pt = apt[1];
               else
                    return;
          }
          else if (mea.Button == MouseButtons.Right)
          {
               if (ModifierKeys == Keys.None)
                    pt = apt[2];
               else if (ModifierKeys == Keys.Shift)
                    pt = apt[3];
               else
                    return;
          }
          else
               return;

          Cursor.Position = PointToScreen(pt);
     }
     protected override void OnMouseMove(MouseEventArgs mea)
     {
          Point pt = new Point(mea.X, mea.Y);

          if (mea.Button == MouseButtons.Left)
          {
               if (ModifierKeys == Keys.Shift)
                    apt[0] = pt;
               else if (ModifierKeys == Keys.None)
                    apt[1] = pt;
               else
                    return;
          }
          else if (mea.Button == MouseButtons.Right)
          {
               if (ModifierKeys == Keys.None)
                    apt[2]= pt;
               else if (ModifierKeys == Keys.Shift)
                    apt[3] = pt;
               else
                    return;
          }
          else
               return;

          Invalidate();
     }
     protected override void OnPaint(PaintEventArgs pea)
     {
          Graphics grfx  = pea.Graphics;
          Brush    brush = new SolidBrush(ForeColor);

          grfx.DrawCurve(new Pen(ForeColor), apt, fTension);

          grfx.DrawString("Tension = " + fTension, Font, brush, 0, 0);

          for (int i = 0; i < 4; i++)
               grfx.FillEllipse(brush, apt[i].X - 3, apt[i].Y - 3, 7, 7);
     }
}
